File: scripts/csv_to_json.py
#!/usr/bin/env python3
"""
CSV to JSON Converter for Sage/Linnworks exports
Usage: python csv_to_json.py input.csv output.json
"""

import csv
import json
import sys
from pathlib import Path

def csv_to_json(csv_file_path, json_file_path):
    """Convert CSV file to JSON format"""
    try:
        data = []
        
        with open(csv_file_path, 'r', encoding='utf-8') as csv_file:
            csv_reader = csv.DictReader(csv_file)
            
            for row in csv_reader:
                # Clean and convert data types
                cleaned_row = {}
                for key, value in row.items():
                    cleaned_row[key.strip()] = clean_value(value)
                data.append(cleaned_row)
        
        # Write JSON file
        with open(json_file_path, 'w', encoding='utf-8') as json_file:
            json.dump(data, json_file, indent=2, ensure_ascii=False)
        
        print(f"✅ Converted {len(data)} records")
        print(f"📄 Input: {csv_file_path}")
        print(f"📄 Output: {json_file_path}")
        
        return data
        
    except Exception as error:
        print(f"❌ Conversion failed: {error}")
        sys.exit(1)

def clean_value(value):
    """Clean and convert value to appropriate type"""
    if not value or value.strip() == '':
        return None
    
    value = value.strip()
    
    # Try to convert to number
    try:
        if '.' in value:
            return float(value)
        else:
            return int(value)
    except ValueError:
        pass
    
    # Convert booleans
    if value.lower() in ['true', 'yes', '1']:
        return True
    if value.lower() in ['false', 'no', '0']:
        return False
    
    # Return as string
    return value

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python csv_to_json.py <input.csv> <output.json>")
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_file = sys.argv[2]
    
    if not Path(input_file).exists():
        print(f"❌ Input file does not exist: {input_file}")
        sys.exit(1)
    
    csv_to_json(input_file, output_file)

File: scripts/csv-to-json.js
#!/usr/bin/env node

/**
 * CSV to JSON Converter for Sage/Linnworks exports
 * Usage: node csv-to-json.js input.csv output.json
 */

const fs = require('fs');
const path = require('path');

function csvToJson(csvFilePath, jsonFilePath) {
  try {
    // Read CSV file
    const csvData = fs.readFileSync(csvFilePath, 'utf8');
    const lines = csvData.split('\n').filter(line => line.trim());

    if (lines.length === 0) {
      throw new Error('CSV file is empty');
    }

    // Parse headers
    const headers = lines[0].split(',').map(header =>
      header.replace(/"/g, '').trim()
    );

    // Parse data rows
    const jsonData = [];
    for (let i = 1; i < lines.length; i++) {
      const values = parseCSVLine(lines[i]);

      if (values.length === headers.length) {
        const obj = {};
        headers.forEach((header, index) => {
          obj[header] = cleanValue(values[index]);
        });
        jsonData.push(obj);
      }
    }

    // Write JSON file
    fs.writeFileSync(jsonFilePath, JSON.stringify(jsonData, null, 2));

    console.log(`✅ Converted ${jsonData.length} records`);
    console.log(`📄 Input: ${csvFilePath}`);
    console.log(`📄 Output: ${jsonFilePath}`);

    return jsonData;

  } catch (error) {
    console.error('❌ Conversion failed:', error.message);
    process.exit(1);
  }
}

function parseCSVLine(line) {
  const values = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];

    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === ',' && !inQuotes) {
      values.push(current.trim());
      current = '';
    } else {
      current += char;
    }
  }

  values.push(current.trim());
  return values;
}

function cleanValue(value) {
  // Remove quotes and clean up data types
  value = value.replace(/^"(.*)"$/, '$1').trim();

  // Convert numbers
  if (/^\d+\.?\d*$/.test(value)) {
    return parseFloat(value);
  }

  // Convert booleans
  if (value.toLowerCase() === 'true') return true;
  if (value.toLowerCase() === 'false') return false;

  // Return string
  return value;
}

// Command line usage
if (require.main === module) {
  const args = process.argv.slice(2);

  if (args.length !== 2) {
    console.log('Usage: node csv-to-json.js <input.csv> <output.json>');
    process.exit(1);
  }

  const [inputFile, outputFile] = args;
  csvToJson(inputFile, outputFile);
}

module.exports = { csvToJson };
